### PR TITLE
editliveos: A full featured replacement for tools/edit-livecd.

### PR DESCRIPTION
**Pull request by FGrose**
```
Edit a LiveOS image to merge a persistent overlay, insert files,
clone a customized instance, adjust the root or home filesystem or
overlay sizes, seclude private or user-specific files, rebuild the
image into a new, .iso image distribution file, and refresh the
source's persistent filesystem overlay. See editliveos --help.

Uses the Device-mapper mirror target to copy the root filesystem,
merging a persistent DM overlay, if present. OverlayFS overlays
are merged by a standard copy of the mounted root filesystem.
```
Note: Python3 was required for lack of hawkey module in Python2
```
    import hawkey
ImportError: No module named hawkey
```
Note also: comment on efdf984

### CODE REVIEW

**Comment by Conan-Kudo** [comment]
@FGrose There is a Python 2 version of the `hawkey` module. It's packaged in Fedora and Mageia as `python2-hawkey`.

File: imgcreate/fs.py [hunk 1]
```diff
@@ -1,8 +1,10 @@
+# coding: utf_8
```
Comment (Conan-Kudo) [inline imgcreate/fs.py]: Please use `utf-8`.

File: imgcreate/editliveos.py [hunk 2]
```diff
@@ -0,0 +1,2161 @@
+#!/usr/bin/python3
```
Comment (Conan-Kudo) [inline imgcreate/editliveos.py]: Please change this to `/usr/bin/python`

File: imgcreate/creator.py [hunk 3]
```diff
@@ -36,14 +37,15 @@
...
 import rpm
 
-from imgcreate.errors import *
-from imgcreate.fs import *
-from imgcreate.dnfinst import *
-from imgcreate import kickstart
+from errors import *
+from fs import *
+from dnfinst import *
+from kickstart import *
```
Comment (Conan-Kudo) [inline imgcreate/creator.py]: Please revert these changes, as they are intended to make it obvious that these modules are internal to livecd-tools.

File: imgcreate/editliveos.py [hunk 4]
```diff
@@ -0,0 +1,2161 @@
...
+import tempfile
+import argparse
+import subprocess
+
+from fs import *
+from live import *
+from debug import *
+from errors import *
+from creator import *
+from kickstart import *
```
Comment (Conan-Kudo) [inline imgcreate/editliveos.py]: Ideally, I'd prefer that all imports be namespaced.

That is, only use `import imgcreate.fs` rather than `from imgcreate.fs import *` and all function calls are called prefixed with the module name.

Since I have not yet changed all the rest of livecd-tools to do this, please at least change from `from fs import *` to `from imgcreate.fs import *`. That way it's easy to tell that these are internal modules and potential conflicts with system modules are avoided.

File: imgcreate/fs.py [hunk 5]
```diff
@@ -16,31 +18,40 @@
...
 import random
 import string
 import logging
 import tempfile
 import time
-from imgcreate.util import call
 
-from imgcreate.errors import *
+from util import *
+from errors import *
```
Comment (Conan-Kudo) [inline imgcreate/fs.py]: Please revert this change.

File: imgcreate/fs.py [hunk 6]
```diff
@@ -65,102 +76,326 @@ def squashfs_compression_type(sqfs_img):
         if p.returncode != 0:
             raise SquashfsError(
                 u"Error while stat-ing '%s'\n'%s'\nreturncode: '%s'" %
-                (args, err, p.returncode))
+                (args, err.decode('utf_8'), p.returncode))
```
Comment (Conan-Kudo) [inline imgcreate/fs.py]: Use `utf-8`.

File: imgcreate/fs.py [hunk 7]
```diff
@@ -285,34 +537,44 @@ def fixed(self):
...
+        losetupProc = subprocess.Popen(['losetup', '-f'],
                                        stdout=subprocess.PIPE)
         losetupOutput = losetupProc.communicate()[0]
 
         if losetupProc.returncode:
             raise MountError("Failed to allocate loop device for '%s'" %
                              self.lofile)
 
-        device = losetupOutput.split()[0].decode("utf-8")
+        device = losetupOutput.split()[0].decode('utf_8')
```
Comment (Conan-Kudo) [inline imgcreate/fs.py]: Use `utf-8`.

File: imgcreate/util.py [hunk 8]
```diff
@@ -37,3 +39,32 @@ def call(*popenargs, **kwargs):
...
+    """Return stdout, stderr, & returncode from a subprocess call."""
+
+    out, err, p, environ = '', '', None, None
+    if env is not None:
+        environ = os.environ.copy()
+        environ.update(env)
+    try:
+        p = subprocess.Popen(args, stdin=subprocess.PIPE, cwd=cwd, env=environ,
+                             stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        out, err = p.communicate(stdin.encode('utf_8'))
```
Comment (Conan-Kudo) [inline imgcreate/util.py]: Use `utf-8`.

File: imgcreate/util.py [hunk 9]
```diff
@@ -37,3 +39,32 @@ def call(*popenargs, **kwargs):
...
+            raise CreatorError('Failed ta execute:\n%s\n'
+                                'error: %s\nstdout: %s\nstderr: %s' %
+                               (args, e, out, err))
+    else:
+        if p.returncode != 0 and raise_err:
+            raise CreatorError('Error in call:\n%s\nenviron: %s\n'
+                                'stdout: %s\nstderr: %s\nreturncode: %s' %
+                               (args, environ, out, err, p.returncode))
+    finally:
+        return out.decode('utf_8'), err.decode('utf_8'), p.returncode
```
Comment (Conan-Kudo) [inline imgcreate/util.py]: Use `utf-8`.

File: imgcreate/editliveos.py [hunk 10]
```diff
@@ -0,0 +1,2161 @@
...
+/^\s*label\s+linux\>/I,/^\s*label\s+check\>/Is/(rd\.live\.image|liveimg).*/\1 quiet/
+/^\s*label\s+check\>/I,/^\s*label\s+vesa\>/Is/(rd\.live\.image|liveimg).*/\1 rd.live.check quiet/
+/^\s*label\s+vesa\>/I,/^\s*label\s+memtest\>/Is/(rd\.live\.image|liveimg).*/\1 nomodeset quiet/
+s/^\s*set\s+timeout=.*/set timeout=60/
+/^\s*menuentry\s+'Start\s+/,/\s+}/{s/(\s+'Start\s+)[^ ]*\s+~/\1/
+s/(rd\.live\.image|liveimg).*/\1 quiet/}
+/^\s*menuentry\s+'Test\s+/,/\s+}/{s/(\s+&\s+start\s+)[^ ]*\s+~/\1/
+s/(rd\.live\.image|liveimg).*/\1 rd.live.check quiet/}
+/^\s*submenu\s+'Trouble/,/\s+}/s/(rd\.live\.image|liveimg).*/\1 nomodeset quiet/
+s/(linuxefi\s+[^ ]+vmlinuz.?)\s+.*\s+(root=live:[^\s+
```
Comment (Conan-Kudo) [inline imgcreate/editliveos.py]: Use `utf-8`.

File: imgcreate/fs.py [hunk 11]
```diff
@@ -16,31 +18,40 @@
...
-from imgcreate.util import call
 
+from imgcreate.util import *
 from imgcreate.errors import *
 
-def makedirs(dirname):
+def chroottypecheck(self, command):
+    """Check for command availability in the chroot image."""
+    with open(os.devnull, 'w') as DEVNULL:
+        return subprocess.call(['type', command], stdout=DEVNULL,
```
Comment (Conan-Kudo) [inline imgcreate/fs.py]: This seems rickety, as it depends on whether `type` actually exists. I'd prefer a Pythonic implementation here.

There's `shutil.which()` in Python 3, and it has [a backport in Python 2](https://pypi.python.org/pypi/backports.shutil_which/3.5.1) ([packaged in Fedora](https://apps.fedoraproject.org/packages/python-backports-shutil_which)).

It'd probably be a good idea to have a wrapper for command checks in chroots and command checks out of chroots to abstract the usage of either `shutil.which()` or the backport.

File: imgcreate/live.py [hunk 12]
```diff
@@ -332,27 +330,26 @@ def __create_iso(self, isodir):
         if subprocess.call(args) != 0:
             raise CreatorError("ISO creation failed!")
 
-        if os.path.exists("/usr/bin/isohybrid"):
+        if subprocess.call(['type', 'isohybrid']):
```
Comment (Conan-Kudo) [inline imgcreate/live.py]: As I mentioned earlier, I'd prefer a Pythonic implementation here.

File: imgcreate/live.py [hunk 13]
```diff
@@ -332,27 +330,26 @@ def __create_iso(self, isodir):
...
-                subprocess.call(["/usr/bin/isohybrid", iso])
+                subprocess.call(["isohybrid", iso])
 
         self.__implant_md5sum(iso)
 
     def __implant_md5sum(self, iso):
         """Implant an isomd5sum."""
-        if os.path.exists("/usr/bin/implantisomd5"):
-            implantisomd5 = "/usr/bin/implantisomd5"
+        if subprocess.call(['type', 'implantisomd5']):
```
Comment (Conan-Kudo) [inline imgcreate/live.py]: As I mentioned earlier, I'd prefer a Pythonic implementation here.

**Comment by FGrose** [comment]
Even more Pythonic is using the **`try:`** **`except:`** construct instead of testing all the likely passes.

**Note:** Testing is needed for the kickstart code, especially.

File: imgcreate/creator.py [hunk 14]
```diff
@@ -1,9 +1,10 @@
 #
 # creator.py : ImageCreator and LoopImageCreator base classes
 #
-# Copyright 2007, Red Hat  Inc.
+# Copyright 2007, Red Hat, Inc.
 # Copyright 2016, Kevin Kofler
 # Copyright 2016, Neal Gompa
+# Copyright 2017, Fedora Project
```
Comment (Conan-Kudo) [inline imgcreate/creator.py]: This is wrong unless you intend to assign rights to the Fedora Project... I think you meant Sugar Labs?

File: imgcreate/debug.py [hunk 15]
```diff
@@ -1,8 +1,9 @@
 #
 # debug.py: Helper routines for debugging
 #
-# Copyright 2008, Red Hat  Inc.
+# Copyright 2008, Red Hat, Inc.
 # Copyright 2016, Neal Gompa
+# Copyright 2017, Fedora Project
```
Comment (Conan-Kudo) [inline imgcreate/debug.py]: Ditto.

File: imgcreate/errors.py [hunk 16]
```diff
@@ -1,7 +1,8 @@
 #
 # errors.py : exception definitions
 #
-# Copyright 2007, Red Hat  Inc.
+# Copyright 2007, Red Hat, Inc.
+# Copyright 2017, Fedora Project
```
Comment (Conan-Kudo) [inline imgcreate/errors.py]: Ditto.

File: imgcreate/live.py [hunk 17]
```diff
@@ -4,6 +4,7 @@
 # Copyright 2007-2012, Red Hat, Inc.
 # Copyright 2016, Kevin Kofler
 # Copyright 2016, Neal Gompa
+# Copyright 2017, Fedora Project
```
Comment (Conan-Kudo) [inline imgcreate/live.py]: Ditto.

File: imgcreate/util.py [hunk 18]
```diff
@@ -1,7 +1,8 @@
 #
 # util.py : Various utility methods
 #
-# Copyright 2010, Red Hat  Inc.
+# Copyright 2010, Red Hat, Inc.
+# Copyright 2017, Fedora Project
```
Comment (Conan-Kudo) [inline imgcreate/util.py]: Ditto.

File: imgcreate/creator.py (hunk 14 above)
Comment (FGrose) [inline imgcreate/creator.py]: It was intentional to assign the rights for the imgcreate components to the Fedora Project.

Comment (Conan-Kudo) [inline imgcreate/creator.py]: Okay, then.

**Comment by Conan-Kudo** [comment]
@FGrose I'm going to try to test this to be sure that it doesn't break building live images. If it doesn't, I'll merge it and cut a v25 release.

**Comment by Conan-Kudo** [comment]
@FGrose Ping! You still interested in making this work?

**Comment by FGrose** [comment]
I rebased and force pushed some fixes for a corner case.

File: imgcreate/kickstart.py [hunk 19]
```diff
@@ -249,12 +266,12 @@ def apply(self, ksrootpw):
...
     """A class to apply a kickstart services configuration to a system."""
     def apply(self, ksservices):
-        if not os.path.exists(self.path("/sbin/chkconfig")):
-            return
-        for s in ksservices.enabled:
-            self.call(["/sbin/chkconfig", s, "on"])
-        for s in ksservices.disabled:
-            self.call(["/sbin/chkconfig", s, "off"])
+
+        if chrootentitycheck('chkconfig', self.instroot):
```
[the remark this answers is not in the document]
Comment (FGrose) [reply imgcreate/kickstart.py]: done.

**Comment by Conan-Kudo** [comment]
Merged, thanks! 💯
